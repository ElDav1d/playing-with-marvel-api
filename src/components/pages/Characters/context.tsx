import { useContext, createContext, Ref } from 'react';

interface Context {
  refProp: Ref<HTMLLIElement> | null;
  counter: number | null;
  setCounter: React.Dispatch<React.SetStateAction<number>> | null;
}

const CharactersContext = createContext<Context>({
  refProp: null,
  counter: null,
  setCounter: null,
});

const useCharactersContext = () => {
  const context = useContext(CharactersContext);

  if (context === undefined) {
    throw new Error('useCharactersContext was used outside of its Provider!');
  }

  return context;
};

export { useCharactersContext };
export default CharactersContext;
