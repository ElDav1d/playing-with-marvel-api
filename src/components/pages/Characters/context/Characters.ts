import { createContext } from 'react';
import combineReducers from 'react-combine-reducers';
import filtersReducer, {
  FiltersState,
  FiltersAction,
  initialFiltersState,
} from '../reducers/FiltersReducer';
import orderReducer, {
  IOrderState,
  IOrderAction,
  initialOrderState,
} from '../reducers/OrderReducer';
import searchStringReducer, {
  SearchStringAction,
  SearchStringState,
  initialSearchStringState,
} from '../reducers/SearchStringReducer';

type Action = SearchStringAction | FiltersAction | IOrderAction;

interface IState {
  searchString: SearchStringState;
  filters: FiltersState;
  order: IOrderState;
}

type CombinedReducer = (state: IState, action: Action) => IState;

type Dispatch = (action: Action) => void;

export const [combinedReducer, initialCombinedState] = combineReducers<CombinedReducer>({
  searchString: [searchStringReducer, initialSearchStringState],
  filters: [filtersReducer, initialFiltersState],
  order: [orderReducer, initialOrderState],
});

const CharactersContext = createContext<
  { charactersContextState: IState; charactersContextDispatch: Dispatch } | undefined
>(undefined);

export default CharactersContext;
