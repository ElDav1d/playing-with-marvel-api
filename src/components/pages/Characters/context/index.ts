export { default as FiltersContext } from './FiltersContext';
export { default as FiltersProvider } from './FiltersProvider';
