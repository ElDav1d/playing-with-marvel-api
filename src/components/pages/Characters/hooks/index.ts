import useCharacters from './useCharacters';
import useDebounce from './useDebounce';

export { useDebounce, useCharacters };
