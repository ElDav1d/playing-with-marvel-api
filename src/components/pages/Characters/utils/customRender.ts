import { render as rtlRender } from '@testing-library/react';
import { ReactElement, JSXElementConstructor } from 'react';

const customRender = (ui: ReactElement<any, string | JSXElementConstructor<any>>, { width = 1024, ...options } = {}) => {
  window.matchMedia = jest.fn().mockImplementation(query => ({
    matches: query.includes(`(min-width: ${width}px)`),
    addListener: jest.fn(),
    removeListener: jest.fn(),
  }));

  return rtlRender(ui, { ...options });
};

export default customRender;
