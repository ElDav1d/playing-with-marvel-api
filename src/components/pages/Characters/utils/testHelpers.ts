// TODO: find a proper way for jest to ignore setup files
/* eslint-disable @typescript-eslint/no-empty-function */
import mockCharactersAZ from '../mocks/mockCharactersAZ.json';
import { useCharacters } from '../hooks';
import userEvent from '@testing-library/user-event';
import { CharacterItem } from '../interfaces/characters';

jest.mock('react-lazy-load-image-component', () => ({
  LazyLoadImage: () => null,
}));

const mockUseCharacters = useCharacters as jest.Mock;

const mockIntersectionObserver = jest.fn();

mockIntersectionObserver.mockReturnValue({
  observe: () => null,
  unobserve: () => null,
  disconnect: () => null,
});

window.IntersectionObserver = mockIntersectionObserver;

export const setUpCharacters = () => JSON.parse(JSON.stringify(mockCharactersAZ));

export const setUpHappyPath = (characters: CharacterItem[]) => {
  mockUseCharacters.mockReturnValue({
    isLoading: false,
    isError: false,
    characters,
  });
};

export const setUpHappyPathWithUser = () => {
  setUpHappyPath(setUpCharacters());
  return { user: userEvent.setup() };
};

export const setUpMatchMedia = (matches: boolean) => {
  Object.defineProperty(window, 'matchMedia', {
    writable: true,
    value: jest.fn().mockImplementation((query) => ({
      matches: matches,
      media: query,
      onchange: null,
      addListener: jest.fn(), // deprecated
      removeListener: jest.fn(), // deprecated
      addEventListener: jest.fn(),
      removeEventListener: jest.fn(),
      dispatchEvent: jest.fn(),
    })),
  });
};
