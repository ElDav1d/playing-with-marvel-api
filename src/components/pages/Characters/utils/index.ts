import customRender from './customRender';
export default customRender