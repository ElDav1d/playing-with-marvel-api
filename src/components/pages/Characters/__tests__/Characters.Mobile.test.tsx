/* eslint-disable @typescript-eslint/no-empty-function */
import { BrowserRouter as Router } from 'react-router-dom';
import { QueryClient, QueryClientProvider } from '@tanstack/react-query';
import { render, screen, waitFor, within } from '@testing-library/react';
import { useCharacters } from '../hooks';
import Characters from '../Characters';
import userEvent from '@testing-library/user-event';
import customRender from '../utils';
import { setUpHappyPathWithUser } from '../utils/testHelpers';

jest.mock('../hooks');

const mockUseCharacters = useCharacters as jest.Mock;

const queryClient = new QueryClient();

jest.mock('react-lazy-load-image-component', () => ({
  LazyLoadImage: () => null,
}));

it('renders the page of characters with the Mobile Characters List Control Panel and matches snapshot', async () => {
  // ARRANGE
  const user = setUpHappyPathWithUser();

  // ACT
  const { asFragment } = render(
    <QueryClientProvider client={queryClient}>
      <Router>
        <Characters />
      </Router>
    </QueryClientProvider>,
  );

  const openButton = screen.getByRole('button', { name: /open/i });

  user.click(openButton);

  await waitFor(() => screen.getByRole('form', { name: /mobile/i }));

  // ASSERT
  expect(asFragment()).toMatchSnapshot();
});

it('opens the Mobile Characters List Control Panel', async () => {
  // ARRANGE
  const user = setUpHappyPathWithUser();

  // ACT
  customRender(
    <QueryClientProvider client={queryClient}>
      <Router>
        <Characters />
      </Router>
    </QueryClientProvider>,
    { width: 500 },
  );

  const openButton = screen.getByRole('button', { name: /open/i });

  user.click(openButton);

  // ASSERT
  await waitFor(() => {
    const mobilePanel = screen.getByRole('form', { name: /mobile/i });
    expect(mobilePanel).toBeInTheDocument();
  });
});

it('closes the Mobile Characters List Control Panel', async () => {
  // ARRANGE
  const user = setUpHappyPathWithUser();

  // ACT
  customRender(
    <QueryClientProvider client={queryClient}>
      <Router>
        <Characters />
      </Router>
    </QueryClientProvider>,
    { width: 500 },
  );

  const openButton = screen.getByRole('button', { name: /open/i });

  user.click(openButton);

  await waitFor(() => {
    const closeButton = screen.getByRole('button', { name: /close/i });

    userEvent.click(closeButton);
  });

  // ASSERT
  await waitFor(() => {
    expect(screen.queryByRole('form', { name: /mobile/i })).toBeNull();
  });
});

it('fetches a new list of characters after typing on search by name input', async () => {
  // ARRANGE
  const user = setUpHappyPathWithUser();

  // TODO: optmise calls
  // At SearchGroup
  // Or avoiding desktop form render
  const DESKTOP_HOOK_CALLS = 1;
  const MOBILE_HOOK_CALLS = 1;
  const INITIAL_HOOK_CALLS = DESKTOP_HOOK_CALLS + MOBILE_HOOK_CALLS;

  // ACT
  customRender(
    <QueryClientProvider client={queryClient}>
      <Router>
        <Characters />
      </Router>
    </QueryClientProvider>,
    { width: 500 },
  );

  const openButton = screen.getByRole('button', { name: /open/i });

  user.click(openButton);

  await waitFor(() => {
    const mobilePanel = screen.getByRole('form', { name: /mobile/i });

    const searchInputGroup = within(mobilePanel).getByRole('group', { name: /search/i });

    const searchInputElement = within(searchInputGroup).getByPlaceholderText(/name/i);

    userEvent.type(searchInputElement, 'X');
  });

  // ASSERT
  await waitFor(() => {
    expect(mockUseCharacters).toHaveBeenCalledTimes(INITIAL_HOOK_CALLS + 1);
    expect(screen.getByRole('heading', { name: /animal/i })).toBeInTheDocument();
    expect(screen.getByRole('heading', { name: /braineater/i })).toBeInTheDocument();
  });
});

it('fetches a new list of characters after selecting an order option', async () => {
  // ARRANGE
  const user = setUpHappyPathWithUser();

  // TODO: optmise calls
  // At SearchGroup
  // Or avoiding CharactersControlPanel isDesktop
  const DESKTOP_HOOK_CALLS = 1;
  const MOBILE_HOOK_CALLS = 1;
  const INITIAL_HOOK_CALLS = DESKTOP_HOOK_CALLS + MOBILE_HOOK_CALLS;

  // ACT
  customRender(
    <QueryClientProvider client={queryClient}>
      <Router>
        <Characters />
      </Router>
    </QueryClientProvider>,
    { width: 500 },
  );

  const openButton = screen.getByRole('button', { name: /open/i });

  user.click(openButton);

  await waitFor(() => {
    const mobilePanel = screen.getByRole('form', { name: /mobile/i });

    const selectInputGroup = within(mobilePanel).getByRole('group', { name: /order/i });

    // reordering not happening because items are mocked
    userEvent.selectOptions(within(selectInputGroup).getByRole('combobox'), '-name');
  });

  // ASSERT
  await waitFor(() => {
    expect(mockUseCharacters).toHaveBeenCalledTimes(INITIAL_HOOK_CALLS + 1);
    expect(screen.getByRole('heading', { name: /animal/i })).toBeInTheDocument();
    expect(screen.getByRole('heading', { name: /braineater/i })).toBeInTheDocument();
  });
});

it('renders an filter input group with its required checkboxes', async () => {
  // ARRANGE
  const user = setUpHappyPathWithUser();

  // ACT
  customRender(
    <QueryClientProvider client={queryClient}>
      <Router>
        <Characters />
      </Router>
    </QueryClientProvider>,
    { width: 500 },
  );

  const openButton = screen.getByRole('button', { name: /open/i });

  user.click(openButton);

  await waitFor(() => {
    const mobilePanel = screen.getByRole('form', { name: /mobile/i });

    const checkInputGroup = within(mobilePanel).getByRole('group', { name: /filter/i });
    const checkWithImage = within(mobilePanel).getByRole('checkbox', { name: /(image)+/i });
    const checkWithDescription = within(mobilePanel).getByRole('checkbox', {
      name: /(description)+/i,
    });

    // ASSERT
    expect(checkInputGroup).toBeInTheDocument();
    expect(checkWithImage).toBeInTheDocument();
    expect(checkWithDescription).toBeInTheDocument();
  });
});

it('renders a characters with image list when the image filter is checked', async () => {
  // ARRANGE
  const user = setUpHappyPathWithUser();

  // ACT
  customRender(
    <QueryClientProvider client={queryClient}>
      <Router>
        <Characters />
      </Router>
    </QueryClientProvider>,
    { width: 500 },
  );

  const openButton = screen.getByRole('button', { name: /open/i });

  user.click(openButton);

  await waitFor(() => {
    const mobilePanel = screen.getByRole('form', { name: /mobile/i });

    const checkInputGroup = within(mobilePanel).getByRole('group', { name: /filter/i });

    const checkWithImage = within(checkInputGroup).getByRole('checkbox', { name: /(image)+/i });

    user.click(checkWithImage);
  });

  // ASSERT
  await waitFor(() => {
    expect(screen.queryByRole('heading', { name: /animal/i })).toBeInTheDocument();
    expect(screen.queryByRole('heading', { name: /crusher/i })).toBeInTheDocument();
    expect(screen.queryByRole('heading', { name: /braineater/i })).toBeNull();
  });
});

it('renders a characters with description list when the description filter is checked', async () => {
  // ARRANGE
  const user = setUpHappyPathWithUser();

  // ACT
  customRender(
    <QueryClientProvider client={queryClient}>
      <Router>
        <Characters />
      </Router>
    </QueryClientProvider>,
    { width: 500 },
  );

  const openButton = screen.getByRole('button', { name: /open/i });

  user.click(openButton);

  await waitFor(() => {
    const mobilePanel = screen.getByRole('form', { name: /mobile/i });

    const checkInputGroup = within(mobilePanel).getByRole('group', { name: /filter/i });

    const checkWithDescription = within(checkInputGroup).getByRole('checkbox', {
      name: /(description)+/i,
    });

    user.click(checkWithDescription);
  });

  // ASSERT
  await waitFor(() => {
    expect(screen.queryByRole('heading', { name: /animal/i })).toBeInTheDocument();
    expect(screen.queryByRole('heading', { name: /braineater/i })).toBeInTheDocument();
    expect(screen.queryByRole('heading', { name: /crusher/i })).toBeNull();
  });
});

it('keeps rendering a filtered characters list after selecting an order option', async () => {
  // ARRANGE
  const user = setUpHappyPathWithUser();

  // ACT
  customRender(
    <QueryClientProvider client={queryClient}>
      <Router>
        <Characters />
      </Router>
    </QueryClientProvider>,
    { width: 500 },
  );

  const openButton = screen.getByRole('button', { name: /open/i });

  user.click(openButton);

  await waitFor(() => {
    const mobilePanel = screen.getByRole('form', { name: /mobile/i });

    const checkInputGroup = within(mobilePanel).getByRole('group', { name: /filter/i });

    const checkWithImage = within(checkInputGroup).getByRole('checkbox', { name: /(image)+/i });

    user.click(checkWithImage);

    const selectInputGroup = within(mobilePanel).getByRole('group', { name: /order/i });

    // reordering not happening because items are mocked
    user.selectOptions(within(selectInputGroup).getByRole('combobox'), '-name');
  });

  // ASSERT
  await waitFor(() => {
    expect(screen.getByRole('heading', { name: /animal/i })).toBeInTheDocument();
    expect(screen.getByRole('heading', { name: /crusher/i })).toBeInTheDocument();
    expect(screen.queryByRole('heading', { name: /braineater/i })).toBeNull();
  });
});

it('renders the complete characters list after unchecking a filter', async () => {
  // ARRANGE
  const user = setUpHappyPathWithUser();

  // ACT
  render(
    <QueryClientProvider client={queryClient}>
      <Router>
        <Characters />
      </Router>
    </QueryClientProvider>,
  );

  const openButton = screen.getByRole('button', { name: /open/i });

  user.click(openButton);

  await waitFor(() => {
    const checkWithImage = screen.getByRole('checkbox', { name: /(image)+/i });

    user.click(checkWithImage);
  });

  // ASSERT
  await waitFor(() => {
    expect(screen.getByRole('heading', { name: /animal/i })).toBeInTheDocument();
    expect(screen.getByRole('heading', { name: /braineater/i })).toBeInTheDocument();
    expect(screen.getByRole('heading', { name: /crusher/i })).toBeInTheDocument();
    expect(screen.getByRole('heading', { name: /destructor/i })).toBeInTheDocument();
  });
});
