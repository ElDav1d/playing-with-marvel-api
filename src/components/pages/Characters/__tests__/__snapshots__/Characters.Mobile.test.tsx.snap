// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders the page of characters with the Mobile Characters List Control Panel and matches snapshot 1`] = `
<DocumentFragment>
  <header
    aria-label="common header"
    class="w-full fixed z-1 top-0 flex justify-center transition-[transform] delay-gridItem duration-header  bg-black "
  >
    <nav
      aria-label="main navigation"
      class="mx-6 md:mx-12 lg:mx-18 xl:mx-auto max-w-[1240px]"
    >
      <a
        class="block focus-visible"
        href="/"
      >
        <h1
          class="text-none"
        >
          Playing with Marvel API
          <svg
            aria-hidden="true"
            height="52"
            role="img"
            width="130"
            xmlns="http://www.w3.org/2000/svg"
          >
            <title>
              MARVEL logo
            </title>
            <rect
              fill="#EC1D24"
              height="100%"
              width="100%"
            />
            <path
              d="M126.222 40.059v7.906H111.58V4h7.885v36.059h6.757zm-62.564-14.5c-.61.294-1.248.44-1.87.442v-14.14h.04c.622-.005 5.264.184 5.264 6.993 0 3.559-1.58 5.804-3.434 6.705zM40.55 34.24l2.183-18.799 2.265 18.799H40.55zm69.655-22.215V4.007H87.879l-3.675 26.779-3.63-26.78h-8.052l.901 7.15c-.928-1.832-4.224-7.15-11.48-7.15-.047-.002-8.06 0-8.06 0l-.031 39.032-5.868-39.031-10.545-.005-6.072 40.44.002-40.435H21.278L17.64 26.724 14.096 4.006H4v43.966h7.95V26.78l3.618 21.192h4.226l3.565-21.192v21.192h15.327l.928-6.762h6.17l.927 6.762 15.047.008h.01v-.008h.02V33.702l1.845-.27 3.817 14.55h7.784l-.002-.01h.022l-5.011-17.048c2.538-1.88 5.406-6.644 4.643-11.203v-.002C74.894 19.777 79.615 48 79.615 48l9.256-.027 6.327-39.85v39.85h15.007v-7.908h-7.124v-10.08h7.124v-8.03h-7.124v-9.931h7.124z"
              fill="#FEFEFE"
            />
            <path
              d="M0 0h30v52H0z"
              fill="#EC1D24"
            />
            <path
              d="M31.5 48V4H21.291l-3.64 22.735L14.102 4H4v44h8V26.792L15.577 48h4.229l3.568-21.208V48z"
              fill="#FEFEFE"
            />
          </svg>
        </h1>
      </a>
    </nav>
    <aside
      class="md:hidden"
    >
      <div
        class="absolute top-0 left-0 h-screen w-full bg-black/50"
        data-testid="dialog-overlay"
      />
      <div
        class="absolute top-0 m-0 h-screen w-[80vw] z-1 px-8 py-11 transition-[left] delay-gridItem duration-gridItem left-0 bg-black"
      >
        <form
          aria-label="Mobile Characters List Control Panel"
          class="focus-within gap-4 flex flex-col"
        >
          <fieldset
            class="text-white"
          >
            <legend
              class="mb-1"
            >
              Search by name
            </legend>
            <input
              class="py-2 px-3 bg-black border border-white shadow appearance-none focus-visible-border w-full md:w-auto"
              placeholder="type a character name"
              type="text"
              value=""
            />
          </fieldset>
          <fieldset
            class=""
          >
            <legend
              class="mb-1"
            >
              Order results
            </legend>
            <div
              class=" css-b62m3t-container"
            >
              <span
                class="css-1f43avz-a11yText-A11yText"
                id="react-select-3-live-region"
              />
              <span
                aria-atomic="false"
                aria-live="polite"
                aria-relevant="additions text"
                class="css-1f43avz-a11yText-A11yText"
                role="log"
              />
              <div
                class=" css-13cymwt-control"
              >
                <div
                  class=" css-1fdsijx-ValueContainer"
                >
                  <div
                    class=" css-1jqq78o-placeholder"
                    id="react-select-3-placeholder"
                  >
                    Select...
                  </div>
                  <div
                    class=" css-qbdosj-Input"
                    data-value=""
                  >
                    <input
                      aria-activedescendant=""
                      aria-autocomplete="list"
                      aria-describedby="react-select-3-placeholder"
                      aria-expanded="false"
                      aria-haspopup="true"
                      autocapitalize="none"
                      autocomplete="off"
                      autocorrect="off"
                      class=""
                      id="react-select-3-input"
                      role="combobox"
                      spellcheck="false"
                      style="opacity: 1; width: 100%; grid-area: 1 / 2; min-width: 2px; border: 0px; margin: 0px; outline: 0; padding: 0px;"
                      tabindex="0"
                      type="text"
                      value=""
                    />
                  </div>
                </div>
                <div
                  class=" css-1hb7zxy-IndicatorsContainer"
                >
                  <span
                    class=" css-1u9des2-indicatorSeparator"
                  />
                  <div
                    aria-hidden="true"
                    class=" css-1xc3v61-indicatorContainer"
                  >
                    <svg
                      aria-hidden="true"
                      class="css-tj5bde-Svg"
                      focusable="false"
                      height="20"
                      viewBox="0 0 20 20"
                      width="20"
                    >
                      <path
                        d="M4.516 7.548c0.436-0.446 1.043-0.481 1.576 0l3.908 3.747 3.908-3.747c0.533-0.481 1.141-0.446 1.574 0 0.436 0.445 0.408 1.197 0 1.615-0.406 0.418-4.695 4.502-4.695 4.502-0.217 0.223-0.502 0.335-0.787 0.335s-0.57-0.112-0.789-0.335c0 0-4.287-4.084-4.695-4.502s-0.436-1.17 0-1.615z"
                      />
                    </svg>
                  </div>
                </div>
              </div>
            </div>
          </fieldset>
          <fieldset
            class="text-white"
          >
            <legend
              class="mb-1"
            >
              Filter results:
            </legend>
            <ul
              class="md:h-full flex flex-col md:flex-row gap-2 md:items-center "
            >
              <li
                class="flex items-center"
              >
                <input
                  aria-label="with image"
                  class="mr-2 h-4 w-4 checked:bg-red bg-black border border-white shadow appearance-none focus-visible-border "
                  id="with image"
                  name="with image"
                  type="checkbox"
                  value="with image"
                />
                <label
                  for="with image"
                >
                  with image
                </label>
              </li>
              <li
                class="flex items-center"
              >
                <input
                  aria-label="with description"
                  class="mr-2 h-4 w-4 checked:bg-red bg-black border border-white shadow appearance-none focus-visible-border "
                  id="with description"
                  name="with description"
                  type="checkbox"
                  value="with description"
                />
                <label
                  for="with description"
                >
                  with description
                </label>
              </li>
            </ul>
          </fieldset>
        </form>
        <button
          aria-label="Close Characters List Control Panel"
          class="absolute top-0 right-0 m-1 p-1 focus-visible"
        >
          <svg
            aria-hidden="true"
            class="h-6 w-6 text-white"
            fill="none"
            stroke="currentColor"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="2"
            viewBox="0 0 24 24"
          >
            <title>
              Close icon
            </title>
            <desc>
              Two crossed strokes
            </desc>
            <line
              x1="18"
              x2="6"
              y1="6"
              y2="18"
            />
            <line
              x1="6"
              x2="18"
              y1="6"
              y2="18"
            />
          </svg>
        </button>
      </div>
    </aside>
  </header>
  <main
    aria-label="characters page main content"
    class="pb-48 min-h-[100vh]"
  >
    <section
      class="relative z-0 flex flex-col items-center justify-around pt-logoDefaultHeight gap-2 min-h-52 md:min-h-72 mb-4 bg-black bg-hero-image bg-center bg-cover before:absolute before:content[\\"\\"] before:h-full before:w-full before:bg-gradient-to-b before:from-trans-0.5-black before:from-50% before:left-0 before:top-0 before:z-[-1]"
    >
      <div
        class="mx-6 md:mx-12 lg:mx-18 xl:mx-auto max-w-[1240px] flex flex-col gap-4"
      >
        <div
          class="w-full text-white text-center"
        >
          <h2
            class="text-2xl md:text-3xl font-semibold text-white text-center uppercase mb-1"
          >
            Marvel Characters
          </h2>
          <p
            class="text-sm md:text-base text-center"
          >
            Get hooked on a hearty helping of heroes and villains from the humble House of Ideas!
          </p>
        </div>
        <form
          aria-label="Desktop Characters List Control Panel"
          class="focus-within gap-4 hidden md:flex justify-center"
        >
          <fieldset
            class="text-white"
          >
            <legend
              class="mb-1"
            >
              Search by name
            </legend>
            <input
              class="py-2 px-3 bg-black border border-white shadow appearance-none focus-visible-border w-full md:w-auto"
              placeholder="type a character name"
              type="text"
              value=""
            />
          </fieldset>
          <fieldset
            class=""
          >
            <legend
              class="mb-1"
            >
              Order results
            </legend>
            <div
              class=" css-b62m3t-container"
            >
              <span
                class="css-1f43avz-a11yText-A11yText"
                id="react-select-2-live-region"
              />
              <span
                aria-atomic="false"
                aria-live="polite"
                aria-relevant="additions text"
                class="css-1f43avz-a11yText-A11yText"
                role="log"
              />
              <div
                class=" css-13cymwt-control"
              >
                <div
                  class=" css-1fdsijx-ValueContainer"
                >
                  <div
                    class=" css-1jqq78o-placeholder"
                    id="react-select-2-placeholder"
                  >
                    Select...
                  </div>
                  <div
                    class=" css-qbdosj-Input"
                    data-value=""
                  >
                    <input
                      aria-activedescendant=""
                      aria-autocomplete="list"
                      aria-describedby="react-select-2-placeholder"
                      aria-expanded="false"
                      aria-haspopup="true"
                      autocapitalize="none"
                      autocomplete="off"
                      autocorrect="off"
                      class=""
                      id="react-select-2-input"
                      role="combobox"
                      spellcheck="false"
                      style="opacity: 1; width: 100%; grid-area: 1 / 2; min-width: 2px; border: 0px; margin: 0px; outline: 0; padding: 0px;"
                      tabindex="0"
                      type="text"
                      value=""
                    />
                  </div>
                </div>
                <div
                  class=" css-1hb7zxy-IndicatorsContainer"
                >
                  <span
                    class=" css-1u9des2-indicatorSeparator"
                  />
                  <div
                    aria-hidden="true"
                    class=" css-1xc3v61-indicatorContainer"
                  >
                    <svg
                      aria-hidden="true"
                      class="css-tj5bde-Svg"
                      focusable="false"
                      height="20"
                      viewBox="0 0 20 20"
                      width="20"
                    >
                      <path
                        d="M4.516 7.548c0.436-0.446 1.043-0.481 1.576 0l3.908 3.747 3.908-3.747c0.533-0.481 1.141-0.446 1.574 0 0.436 0.445 0.408 1.197 0 1.615-0.406 0.418-4.695 4.502-4.695 4.502-0.217 0.223-0.502 0.335-0.787 0.335s-0.57-0.112-0.789-0.335c0 0-4.287-4.084-4.695-4.502s-0.436-1.17 0-1.615z"
                      />
                    </svg>
                  </div>
                </div>
              </div>
            </div>
          </fieldset>
          <fieldset
            class="text-white"
          >
            <legend
              class="mb-1"
            >
              Filter results:
            </legend>
            <ul
              class="md:h-full flex flex-col md:flex-row gap-2 md:items-center "
            >
              <li
                class="flex items-center"
              >
                <input
                  aria-label="with image"
                  class="mr-2 h-4 w-4 checked:bg-red bg-black border border-white shadow appearance-none focus-visible-border "
                  id="with image"
                  name="with image"
                  type="checkbox"
                  value="with image"
                />
                <label
                  for="with image"
                >
                  with image
                </label>
              </li>
              <li
                class="flex items-center"
              >
                <input
                  aria-label="with description"
                  class="mr-2 h-4 w-4 checked:bg-red bg-black border border-white shadow appearance-none focus-visible-border "
                  id="with description"
                  name="with description"
                  type="checkbox"
                  value="with description"
                />
                <label
                  for="with description"
                >
                  with description
                </label>
              </li>
            </ul>
          </fieldset>
        </form>
      </div>
    </section>
    <section
      class="mx-6 md:mx-12 lg:mx-18 xl:mx-auto max-w-[1240px]"
    >
      <ul
        aria-live="polite"
        class="grid gap-3 grid-flow-row grid-cols-auto-min-max-120-auto md:grid-cols-auto-min-max-185-auto"
      >
        <li
          class="bg-black text-white overflow-hidden relative z-0
      after:absolute after:content[\\"\\"] after:z-1 after:bottom-0 after:right-0 after:border-8 after:border-t-transparent after:border-r-white after:border-b-white after:border-l-transparent
      
    "
        >
          <a
            aria-labelledby="1011176"
            class="block group"
            href="/character/1011176/animal"
          >
            <picture
              class="relative after:absolute after:content[\\"\\"] after:bg-red after:h-1 after:w-full after:left-0 after:-bottom-0.5"
            />
            <div
              class="relative z-0 h-40 p-3 before:absolute before:transition-[max-height] before:ease-in-out before:delay-gridItem before:duration-gridItem before:content[\\"\\"] before:h-[105%] before:max-h-[0px] before:bg-red before:w-full before:left-0 before:-top-0.5 before:z-[-1] group-hover:before:max-h-[300px]"
            >
              <h3
                class="mb-2 font-semibold uppercase text-sm leading-5 line-clamp-2"
              >
                Animal
              </h3>
              <p
                class="text-xs leading-5 line-clamp-4"
              >
                Xorn, whose mutation manifested as a tiny, powerful star inside his head, was brainwashed into believing he was the reincarnation of Magneto.
              </p>
            </div>
          </a>
        </li>
        <li
          class="bg-black text-white overflow-hidden relative z-0
      after:absolute after:content[\\"\\"] after:z-1 after:bottom-0 after:right-0 after:border-8 after:border-t-transparent after:border-r-white after:border-b-white after:border-l-transparent
      
    "
        >
          <a
            aria-labelledby="1011177"
            class="block group"
            href="/character/1011177/braineater"
          >
            <picture
              class="relative after:absolute after:content[\\"\\"] after:bg-red after:h-1 after:w-full after:left-0 after:-bottom-0.5"
            />
            <div
              class="relative z-0 h-40 p-3 before:absolute before:transition-[max-height] before:ease-in-out before:delay-gridItem before:duration-gridItem before:content[\\"\\"] before:h-[105%] before:max-h-[0px] before:bg-red before:w-full before:left-0 before:-top-0.5 before:z-[-1] group-hover:before:max-h-[300px]"
            >
              <h3
                class="mb-2 font-semibold uppercase text-sm leading-5 line-clamp-2"
              >
                Braineater
              </h3>
              <p
                class="text-xs leading-5 line-clamp-4"
              >
                Xorn, whose mutation manifested as a tiny, powerful star inside his head, was brainwashed into believing he was the reincarnation of Magneto.
              </p>
            </div>
          </a>
        </li>
        <li
          class="bg-black text-white overflow-hidden relative z-0
      after:absolute after:content[\\"\\"] after:z-1 after:bottom-0 after:right-0 after:border-8 after:border-t-transparent after:border-r-white after:border-b-white after:border-l-transparent
      
    "
        >
          <a
            aria-labelledby="1011178"
            class="block group"
            href="/character/1011178/crusher"
          >
            <picture
              class="relative after:absolute after:content[\\"\\"] after:bg-red after:h-1 after:w-full after:left-0 after:-bottom-0.5"
            />
            <div
              class="relative z-0 h-40 p-3 before:absolute before:transition-[max-height] before:ease-in-out before:delay-gridItem before:duration-gridItem before:content[\\"\\"] before:h-[105%] before:max-h-[0px] before:bg-red before:w-full before:left-0 before:-top-0.5 before:z-[-1] group-hover:before:max-h-[300px]"
            >
              <h3
                class="mb-2 font-semibold uppercase text-sm leading-5 line-clamp-2"
              >
                Crusher
              </h3>
            </div>
          </a>
        </li>
        <li
          class="bg-black text-white overflow-hidden relative z-0
      after:absolute after:content[\\"\\"] after:z-1 after:bottom-0 after:right-0 after:border-8 after:border-t-transparent after:border-r-white after:border-b-white after:border-l-transparent
      
    "
        >
          <a
            aria-labelledby="1011179"
            class="block group"
            href="/character/1011179/destructor"
          >
            <picture
              class="relative after:absolute after:content[\\"\\"] after:bg-red after:h-1 after:w-full after:left-0 after:-bottom-0.5"
            />
            <div
              class="relative z-0 h-40 p-3 before:absolute before:transition-[max-height] before:ease-in-out before:delay-gridItem before:duration-gridItem before:content[\\"\\"] before:h-[105%] before:max-h-[0px] before:bg-red before:w-full before:left-0 before:-top-0.5 before:z-[-1] group-hover:before:max-h-[300px]"
            >
              <h3
                class="mb-2 font-semibold uppercase text-sm leading-5 line-clamp-2"
              >
                Destructor
              </h3>
            </div>
          </a>
        </li>
      </ul>
    </section>
  </main>
  <footer
    aria-label="common footer"
    class="absolute bottom-0 w-full p-6 bg-stone-900"
  >
    <section
      class="mx-6 md:mx-12 lg:mx-18 xl:mx-auto max-w-[1240px]"
    >
      <a
        class="block active focus-visible"
        href="/"
      >
        <span
          aria-hidden="true"
        >
          <svg
            aria-hidden="true"
            height="115"
            viewBox="0 0 36 52"
            width="80"
            xmlns="http://www.w3.org/2000/svg"
          >
            <title>
              MARVEL only M logo
            </title>
            <rect
              fill="#EC1D24"
              height="100%"
              width="100%"
            />
            <path
              d="M31.5 48V4H21.291l-3.64 22.735L14.102 4H4v44h8V26.792L15.577 48h4.229l3.568-21.208V48z"
              fill="#FEFEFE"
            />
          </svg>
        </span>
      </a>
    </section>
  </footer>
</DocumentFragment>
`;
